ecrtokenrefresher/pkg/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,142 @@
+package k8s
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"os"
+	"path/filepath"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type config struct {
+	Auths map[string]*auth `json:"auths"`
+}
+
+type auth struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+	Email    string `json:"email"`
+	Auth     string `json:"auth"`
+}
+
+const (
+	defaultEmail      = "test@test.com"
+	configMapName     = "ns-secret-map"
+	packagesNamespace = "eksa-packages"
+)
+
+func UpdatePasswords(name string, username string, token string, registries string) (error, []string) {
+	failedList := make([]string, 0)
+	clientset, err := getClientSet()
+	if err != nil {
+		return err, failedList
+	}
+
+	ecrAuth, err := createECRAuthConfig(username, token, registries)
+	if err != nil {
+		return err, failedList
+	}
+
+	targetNamespaces, err := getNamespacesFromConfigMap(clientset)
+	if err != nil {
+		return err, failedList
+	}
+
+	for _, ns := range targetNamespaces {
+		secret, err := getSecret(clientset, name, ns)
+		if secret == nil {
+			secret = createSecret(name, ns)
+			secret.Data[corev1.DockerConfigJsonKey] = ecrAuth
+			_, err = clientset.CoreV1().Secrets(ns).Create(context.TODO(), secret, metav1.CreateOptions{})
+			if err != nil {
+				failedList = append(failedList, ns)
+			}
+		} else {
+			secret.Data[corev1.DockerConfigJsonKey] = ecrAuth
+			_, err = clientset.CoreV1().Secrets(ns).Update(context.TODO(), secret, metav1.UpdateOptions{})
+			if err != nil {
+				failedList = append(failedList, ns)
+			}
+		}
+	}
+
+	return err, failedList
+}
+
+func getClientSet() (*kubernetes.Clientset, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		// Program is not being run from inside cluster. Try default kubeconfig
+		config, err = clientcmd.BuildConfigFromFlags("",
+			filepath.Join(os.Getenv("HOME"), ".kube", "config"))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, err
+}
+
```

Comment:
I could wrap my head around this a bit easier with an OO design. If everything was organized by noun, the noun here being secret. We'd have some secret object that has all these methods as receivers. I think we can live with this for now since it is just temporary I hope.